### PR TITLE
Add Node v18 support and drop Node v12

### PR DESCRIPTION
**Pull request by davidmurdoch**
Addresses #5498  

This PR drops Node v12 and adds Node 18 support. This is accomplished by fixing 1) truffle-db's reliance on an unsupported dependency; 2) eliminating a transitive `hdwallet-provider` dependency incompatible with the OpenSSL 3.0 which Node introduced in version 17+; 3) Introduce a temporary `truffle/hdwallet` package that will eventually be replaced by a `@ganache` package.

## truffle-db
- [x] remove `pouchdb-adapter-node-websql` and rely on `indexedDB` for database persistence
- [x] set the default storage directory to end in `indexeddb`. This will bypass mime type checks, and we should worry about reading sqlite3 files as `indexedDB` files.
- [x] when`truffle-config` specifies a `db.adapter.sqlite` configuration, it will be ignored and `indexeddb` will be used. A message will be issued to the user informing them.

### Data store upgradable concerns

These concerns are not addressed in this PR, as truffle-db is an experimental feature not enabled by default. 
  1. write archiving functionality to back up existing user data. 
  1. document path forward to actually use said archived data. See: https://pouchdb.com/2018/01/23/pouchdb-6.4.2.html
  1. db is no longer concurrent. indexedDB lock mechanism can only support one read & write access. This means that we cannot use `db serve` and `migrate | compile` at the same time.

## hdwallet-provider

`hdwallet-provider`'s had a transitive dependency that's incompatible with node's OpenSSL.
- [x] drop `etherumjs-wallet` which is incompatible with OpenSSL 3.0 because it relies on algorithms and/or key sizes no longer supported by default. [See Node release blog for more details](https://nodejs.org/es/blog/release/v17.0.0/#openssl-3-0)
- [x] introduce `@truffle/hdwallet` to handle account generation from seed and hd-path. Please scrutinize!

# How to test

1. test with truffle db enabled in a truffle project (`MetaCoin`)
1. verify db files created are of `indexedDB` format (they look like `LevelDB`)

### CODE REVIEW

File: packages/db/src/meta/pouch/adapters/index.ts
Comment (cds-amal) [inline packages/db/src/meta/pouch/adapters/index.ts:80]: There could be projects that defined a sqlite adapter. Without this case, the truffle command will throw in default handler. Is this the right way to proceed? I'm not sure, so please chime in, @gnidan @davidmurdoch.

File: packages/db/src/meta/pouch/adapters/indexeddb.ts
Comment (cds-amal) [inline packages/db/src/meta/pouch/adapters/indexeddb.ts:24]: NOTE: This is to makes sure that the new default persistence location (indexeddb) does not collide with previous default location (sqlite)

File: packages/db/src/meta/pouch/adapters/index.ts
Comment (davidmurdoch) [reply packages/db/src/meta/pouch/adapters/index.ts:80]: The only change I would make is change the word "Deprecated", since it's been removed altogether.

**Comment by haltman-at** [comment]
Note: Apparently this is ready for review despite the unchecked checkboxes.

**Comment by eggplantzzz** [comment]
Ignore the unchecked checkboxes, this is ready for review.

File: packages/db/src/meta/pouch/adapters/indexeddb.ts
Comment (eggplantzzz) [inline packages/db/src/meta/pouch/adapters/indexeddb.ts:30]: It uses `indexeddb` by default? Should we specify it here for clarity? Maybe the comment is enough...

File: packages/hdwallet/package.json
Comment (eggplantzzz) [inline packages/hdwallet/package.json:1]: Would it be better to name this something more general like @truffle/cryptography? Not something that should hold this PR back, just asking.

Comment (davidmurdoch) [reply packages/hdwallet/package.json:1]: Eh, I don't think so. This package will likely be removed from truffle and moved over to Ganache eventually (these methods were actually pulled from Ganache's source code, so it'd make sense for ganache to provide a package as a home for them).

**Review by haltman-at — COMMENTED** [review]
(Partial review since I'm too tired right now to do a full review.)

**Review by lsqproduction — COMMENTED** [review]
I'm also getting the error ` /bin/sh: line 0: exit: too many arguments` on ` yarn test`.

**Comment by cds-amal** [comment]
> I'm also getting the error ` /bin/sh: line 0: exit: too many arguments` on ` yarn test`.

@lsqproduction, was `yarn test` invoked from `truffle root`, or `package/db`? I tested this from `package/db` with both node 16 and 18 LTS.  Did you bootstrap?

**Comment by lsqproduction** [comment]
> > I'm also getting the error ` /bin/sh: line 0: exit: too many arguments` on ` yarn test`.
> 
> @lsqproduction, was `yarn test` invoked from `truffle root`, or `package/db`? 

I tested this from `package/db` with both node 16 and 18 LTS. Did you bootstrap?
Hi @cds-amal, I invoked `yarn test` from `truffle root`, the error came up on `@truffle/hdwallet`. 
It is passing from `package/db`.

**Comment by cds-amal** [comment]
>I invoked yarn test from truffle root, the error came up on @truffle/hdwallet.
>It is passing from package/db.

Is this for Node 16, or Node 18?

**Comment by lsqproduction** [comment]
> > I invoked yarn test from truffle root, the error came up on @truffle/hdwallet.
> > It is passing from package/db.
> 
> Is this for Node 16, or Node 18?

node -v                                                                    1 ↵
v18.4.0

File: packages/hdwallet/package.json
Comment (cds-amal) [inline packages/hdwallet/package.json:20]: @lsqproduction discovered an issue with Node 18, on M1 OS X which is triggered by running `yarn test` from the repo root. This can be simulated by scoping `@truffle/hdwallet` via `$(yarn bin)/lerna run test --stream --concurrency=1 --scope @truffle/hdwallet -- --colors`. 

One suggestion is to remove the `test` script from `@truffle/hdwallet` as the test script does nothing.

# error output
```
$ $(yarn bin)/lerna run test --stream --concurrency=1 --scope @truffle/hdwallet -- --colors
lerna notice cli v4.0.0
lerna info versioning independent
lerna notice filter including "@truffle/hdwallet"
lerna info filter [ '@truffle/hdwallet' ]
lerna info Executing command in 1 package: "yarn run test --colors"
yarn run v1.22.19
$ exit 0 --colors
@truffle/hdwallet: /bin/sh: line 0: exit: too many arguments
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
lerna ERR! yarn run test --colors exited 1 in '@truffle/hdwallet'
```

**Review by cds-amal — COMMENTED** [review]
Documenting @lsqproduction's discovery.
M1, OS X (unknown), Node v18.4.0, confirmed in Node v18.9.1

File: packages/hdwallet-provider/src/index.ts
Comment (haltman-at) [inline packages/hdwallet-provider/src/index.ts:330]: So, this line makes me think that the function `createAccountGeneratorFromSeedAndPath` is badly abstracted.

If it took a string, and performed the `split("/")` itself, that would make sense.  If, by contrast, it simply took an array of number/boolean pairs, and didn't perform any string processing, that would also make sense.  As it is, though, the level of abstraction is somewhere inbetween, which makes no sense.  It *doesn't* perform the split, but it *does* check for the initial m/M, etc?

It would probably make more sense to break this up a bit.  Have the publicly exposed function in `@truflfe/hdwallet` take a string, not an array.  It could then perform the split, check for `m`, possibly perform other validation (currently there mostly isn't any, like I could pass in "m/1potato"), check for apostrophes, turn it all into a nice list of number/boolean pairs, and then pass that to another function that did the rest of the work.  That, to my mind, would make much more sense in terms of each function representing a sensible abstraction.

**Review by haltman-at — COMMENTED** [review]
OK, back with a bunch of style comments on `@truffle/hdwallet`, in terms of how to make it clearer!  Also I think one of the functions is at the wrong level of abstraction.

File: packages/hdwallet/src/index.ts
Comment (davidmurdoch) [reply packages/hdwallet/src/index.ts:29]: Monomorphic fns are faster 🙂

File: packages/hdwallet-provider/src/index.ts
Comment (davidmurdoch) [reply packages/hdwallet-provider/src/index.ts:336]: Buffer.from doesn't copy.  IIRC, the value is a Unit8Array which doesn't have toString("hex")

File: packages/hdwallet-provider/src/index.ts
Comment (davidmurdoch) [reply packages/hdwallet-provider/src/index.ts:330]: This code is from ganache, where it makes a bit more sense. In any case, the idea is to move this package to ganache, where it originated, but we didn't want to wait a release cycle.

Ganache tries to avoid redundant validation for internal packages, so if validation is added here, we likely won't pull it back into Ganache. In other words: don't send these functions nonsense. 🤪

Comment (davidmurdoch) [reply packages/hdwallet-provider/src/index.ts:330]: (I do agree that it's a strange function)

File: packages/hdwallet/src/index.ts
Comment (davidmurdoch) [reply packages/hdwallet/src/index.ts:11]: What are you, cave people? What do you use?

File: packages/hdwallet/src/index.ts
Comment (davidmurdoch) [reply packages/hdwallet/src/index.ts:119]: It's slower.

File: packages/hdwallet/src/index.ts
Comment (davidmurdoch) [reply packages/hdwallet/src/index.ts:120]: It's for performance.  We already know the length,. So why make Buffer figure it out internally, which it does by looping over the input array?

File: packages/hdwallet/src/index.ts
Comment (davidmurdoch) [reply packages/hdwallet/src/index.ts:127]: I don't know either. This stuff was lifted from some crypto modules. 😅

File: packages/hdwallet/src/index.ts
Comment (davidmurdoch) [reply packages/hdwallet/src/index.ts:131]: Performance

File: packages/hdwallet/src/index.ts
Comment (davidmurdoch) [reply packages/hdwallet/src/index.ts:135]: IIRC, because it is a Uint8Array.  And it's not a copy.

File: packages/hdwallet/src/index.ts
Comment (haltman-at) [reply packages/hdwallet/src/index.ts:119]: Really? Huh! That's surprising...

File: packages/hdwallet/src/index.ts
Comment (haltman-at) [reply packages/hdwallet/src/index.ts:120]: Wait, really, it does that?  Yikes!  Might want to add a comment about that so someone doesn't remove it in the future...

File: packages/hdwallet/src/index.ts
Comment (haltman-at) [reply packages/hdwallet/src/index.ts:127]: Haha, OK, guess it's fine then. XD

File: packages/hdwallet/src/index.ts
Comment (haltman-at) [reply packages/hdwallet/src/index.ts:131]: Huh, that makes a difference?  OK.

File: packages/hdwallet-provider/src/index.ts
Comment (haltman-at) [reply packages/hdwallet-provider/src/index.ts:336]: [Buffer.from(buffer)](https://nodejs.org/docs/latest-v14.x/api/buffer.html#buffer_static_method_buffer_from_buffer) does copy, according to the docs.  But if it's a `Uint8Array` then this makes sense!

File: packages/hdwallet/src/index.ts
Comment (haltman-at) [reply packages/hdwallet/src/index.ts:135]: [Buffer.from(buffer)](https://nodejs.org/docs/latest-v14.x/api/buffer.html#buffer_static_method_buffer_from_buffer) does copy, according to the docs.  But if it's a `Uint8Array` then this makes sense!

File: packages/hdwallet-provider/src/index.ts
Comment (haltman-at) [reply packages/hdwallet-provider/src/index.ts:330]: Oh, ha.  I assumed they were taking user input.  OK, do it your way then I guess!

File: packages/hdwallet/src/index.ts
Comment (haltman-at) [reply packages/hdwallet/src/index.ts:11]: They just get named in camel case like everything else. :shrug:

File: packages/hdwallet/src/index.ts
Comment (haltman-at) [reply packages/hdwallet/src/index.ts:29]: Well now you're making me wonder if you should be specifying the rest of the arguments as well in that case. :P

File: packages/db/src/meta/pouch/adapters/indexeddb.ts
Comment (cds-amal) [reply packages/db/src/meta/pouch/adapters/indexeddb.ts:30]: I think the comment conveys that, but am happy to change it to be more specific is you wish.

File: packages/db/src/test/persistence.spec.ts
Comment (cds-amal) [reply packages/db/src/test/persistence.spec.ts:53]: It seems like it should.

File: packages/db/ARCHITECTURE.md
Comment (cds-amal) [reply packages/db/ARCHITECTURE.md:41]: I don't think so, that website is a good landing page.

File: packages/hdwallet/src/index.ts
Comment (cds-amal) [reply packages/hdwallet/src/index.ts:52]: Oh yea, `{1}` is not needed here. @haltman-at, what's the suggested use of `$`?

File: packages/hdwallet/src/index.ts
Comment (davidmurdoch) [reply packages/hdwallet/src/index.ts:135]: Ah, sorry. I stand corrected. I think that ideally we wouldn't copy here, as you suggested.

File: packages/hdwallet/src/index.ts
Comment (haltman-at) [reply packages/hdwallet/src/index.ts:52]: Well I thought they might be trying to check specifically for the strings "m" or "M", rather than anything beginning with those.

File: packages/hdwallet/src/index.ts
Comment (haltman-at) [reply packages/hdwallet/src/index.ts:127]: Although, thought -- maybe we should put in comments crediting the crypto modules so we don't lose track of the origin here?

File: packages/hdwallet-provider/src/index.ts
Comment (gnidan) [reply packages/hdwallet-provider/src/index.ts:51]: As a result of this change, will it be able to access the private key(s) via object member access?

Like does `provider.wallets` contain private keys?

Comment (gnidan) [reply packages/hdwallet-provider/src/index.ts:51]: I guess `wallets` here has been exposing private keys for some time.

We have to fix this.

**Review by gnidan — DISMISSED** [review]
Requesting changes because we can't be allowing HDWalletProvider to expose private keys at all.

File: packages/core/package.json
Comment (cds-amal) [reply packages/core/package.json:43]: It's to set the version correctly for the release process.

**Comment by cds-amal** [comment]
Responding to [@davidmurdoch 's comment above](https://github.com/trufflesuite/truffle/pull/5543#discussion_r983057237)
> or wait until this package is moved to Ganache (if it actually gets moved!).

I think we should decide if this is a temporary module to eventually move to Ganache, or, it's a permanent addition to Truffle.

File: packages/db/src/meta/pouch/adapters/indexeddb.ts
Comment (eggplantzzz) [reply packages/db/src/meta/pouch/adapters/indexeddb.ts:24]: Why did this get moved here from line 15?

File: packages/core/package.json
Comment (davidmurdoch) [reply packages/core/package.json:43]: Ah, for pre-release. Got it.

File: packages/hdwallet-provider/src/index.ts
Comment (davidmurdoch) [reply packages/hdwallet-provider/src/index.ts:51]: Just to clear this conversation, we _don't_ need to do this in this PR, right?

File: packages/db/src/meta/pouch/adapters/indexeddb.ts
Comment (cds-amal) [reply packages/db/src/meta/pouch/adapters/indexeddb.ts:24]: This is to make sure that the persistence file path for the default `indexeddb` will always end in `indexeddb` to prevent collision with earlier versions that rely solely on the default path at line 15. This should hold true for user-level, or per-project configs. Should I update the comment?

**Comment by cds-amal** [comment]
Ack, rebasing to unblock CI

File: packages/core/lib/commands/init/initSource/truffle-config.js
Comment (cds-amal) [reply packages/core/lib/commands/init/initSource/truffle-config.js:109]: I'm going to skip this and live w/ prettier's opinion for now.

File: packages/hdwallet/package.json
Comment (cds-amal) [reply packages/hdwallet/package.json]: This issue was caused by `hdwallet` not properly invoking its tests. It has been [fixed in this commit](https://github.com/trufflesuite/truffle/pull/5543/commits/7769645ba4c7d4fcd3f4a0262cbccf38a9b127c2). Thanks for bringing attention to this issue @lsqproduction. We all missed it!

File: .github/workflows/nodejs.yml
Comment (eggplantzzz) [inline .github/workflows/nodejs.yml:52]: do we not need to use npm 6 any more? are the more recent version on npm AOK now?

File: packages/db/src/meta/pouch/adapters/indexeddb.ts
Comment (eggplantzzz) [inline packages/db/src/meta/pouch/adapters/indexeddb.ts:24]: why did adding this suffix get moved? is it just more proper like this since default settings should probably be a little more like, "general"?

**Review by eggplantzzz — APPROVED** [review]
me like

File: packages/db/src/meta/pouch/adapters/indexeddb.ts
Comment (cds-amal) [reply packages/db/src/meta/pouch/adapters/indexeddb.ts:24]: This is to ensure the new truffle db persistence location doesn't conflict with existing (default and customized) locations.

File: packages/hdwallet/src/index.ts
Comment (davidmurdoch) [reply packages/hdwallet/src/index.ts:135]: Note: TIL that `Buffer.from(newPrivK.buffer)` isn't quite enough, we'd need `Buffer.from(newPrivK.buffer, newPrivK.byteOffset, newPrivK.byteLength)`

**Pull request merged**
